// @jbrowse/plugin-linear-genome-view/esm/LinearGenomeView/components/LinearGenomeView.js
import React, { lazy, Suspense, useEffect, useRef } from 'react';
import { makeStyles } from 'tss-react/mui';
import { LoadingEllipses } from '@jbrowse/core/ui';
import { getSession } from '@jbrowse/core/util';
import { observer } from 'mobx-react';
import TrackContainer from './TrackContainer';
import TracksContainer from './TracksContainer';
const ImportForm = lazy(() => import('./ImportForm'));
const NoTracksActiveButton = lazy(() => import('./NoTracksActiveButton'));
const useStyles = makeStyles()(theme => ({
    note: {
        textAlign: 'center',
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    rel: {
        position: 'relative',
    },
    top: {
        zIndex: 1000,
    },
}));
const LinearGenomeView = observer(function ({ model, }) {
    const { tracks, error, initialized, hasDisplayedRegions } = model;
    const ref = useRef(null);
    const session = getSession(model);
    const { classes } = useStyles();
    useEffect(() => {
        // sets the focused view id based on a click within the LGV;
        // necessary for subviews to be focused properly
        function handleSelectView(e) {
            var _a, _b;
            if (e.target instanceof Element && ((_a = ref.current) === null || _a === void 0 ? void 0 : _a.contains(e.target))) {
                (_b = session.setFocusedViewId) === null || _b === void 0 ? void 0 : _b.call(session, model.id);
            }
        }
        document.addEventListener('mousedown', handleSelectView);
        document.addEventListener('keydown', handleSelectView);
        return () => {
            document.removeEventListener('mousedown', handleSelectView);
            document.removeEventListener('keydown', handleSelectView);
        };
    }, [session, model]);
    if (!initialized && !error) {
        return React.createElement(LoadingEllipses, { variant: "h6" });
    }
    if (!hasDisplayedRegions || error) {
        return React.createElement(ImportForm, { model: model });
    }
    const MiniControlsComponent = model.MiniControlsComponent();
    const HeaderComponent = model.HeaderComponent();
    return (React.createElement("div", { className: classes.rel, ref: ref, onMouseLeave: () => {
            session.setHovered(undefined);
        }, onMouseMove: event => {
            const c = ref.current;
            if (!c) {
                return;
            }
            const { tracks } = model;
            const leftPx = event.clientX - c.getBoundingClientRect().left;
            const hoverPosition = model.pxToBp(leftPx);
            const hoverFeature = tracks.find(t => t.displays[0].featureUnderMouse);
            session.setHovered({ hoverPosition, hoverFeature });
        } },
        React.createElement(HeaderComponent, { model: model }),
        React.createElement(MiniControlsComponent, { model: model }),
        React.createElement(TracksContainer, { model: model }, !tracks.length ? (React.createElement(Suspense, { fallback: React.createElement(React.Fragment, null) },
            React.createElement(NoTracksActiveButton, { model: model }))) : (tracks.map(track => (React.createElement(TrackContainer, { key: track.id, model: model, track: track })))))));
});
export default LinearGenomeView;
