// @jbrowse/plugin-linear-genome-view/esm/LinearGenomeView/components/NoTracksActiveButton.js
import React from 'react';
import { Button, Paper, Typography } from '@mui/material';
import { makeStyles } from 'tss-react/mui';
import { observer } from 'mobx-react';
// icons
import { TrackSelector as TrackSelectorIcon } from '@jbrowse/core/ui/Icons';
const useStyles = makeStyles()(theme => ({
    note: {
        textAlign: 'center',
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    top: {
        zIndex: 1000,
    },
}));
const NoTracksActiveButton = observer(function ({ model, }) {
    const { classes } = useStyles();
    const { hideNoTracksActive } = model;
    return (React.createElement(Paper, { className: classes.note }, !hideNoTracksActive ? (React.createElement(React.Fragment, null,
        React.createElement(Typography, null, "No tracks active."),
        React.createElement(Button, { variant: "contained", color: "primary", onClick: () => model.activateTrackSelector(), className: classes.top, startIcon: React.createElement(TrackSelectorIcon, null) }, "Open track selector"))) : (React.createElement("div", { style: { height: '48px' } }))));
});
export default NoTracksActiveButton;
