// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/NoTracksActiveButton.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
// icons
const Icons_1 = require("@jbrowse/core/ui/Icons");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    note: {
        textAlign: 'center',
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    top: {
        zIndex: 1000,
    },
}));
const NoTracksActiveButton = (0, mobx_react_1.observer)(function ({ model, }) {
    const { classes } = useStyles();
    const { hideNoTracksActive } = model;
    return (react_1.default.createElement(material_1.Paper, { className: classes.note }, !hideNoTracksActive ? (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(material_1.Typography, null, "No tracks active."),
        react_1.default.createElement(material_1.Button, { variant: "contained", color: "primary", onClick: () => model.activateTrackSelector(), className: classes.top, startIcon: react_1.default.createElement(Icons_1.TrackSelector, null) }, "Open track selector"))) : (react_1.default.createElement("div", { style: { height: '48px' } }))));
});
exports.default = NoTracksActiveButton;
