// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/OverviewRubberband.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const util_1 = require("@jbrowse/core/util");
const mobx_react_1 = require("mobx-react");
const RubberbandSpan_1 = __importDefault(require("./RubberbandSpan"));
const util_2 = require("./util");
const useStyles = (0, mui_1.makeStyles)()({
    rubberbandControl: {
        cursor: 'crosshair',
        width: '100%',
        minHeight: 8,
    },
    guide: {
        pointerEvents: 'none',
        height: '100%',
        width: 1,
        position: 'absolute',
        zIndex: 10,
    },
    rel: {
        position: 'relative',
    },
});
const HoverTooltip = (0, mobx_react_1.observer)(function ({ model, open, guideX, overview, }) {
    var _a;
    const { classes } = useStyles();
    const { cytobandOffset } = model;
    const { assemblyManager } = (0, util_1.getSession)(model);
    const px = overview.pxToBp(guideX - cytobandOffset);
    const assembly = assemblyManager.get(px.assemblyName);
    const cytoband = (_a = assembly === null || assembly === void 0 ? void 0 : assembly.cytobands) === null || _a === void 0 ? void 0 : _a.find(f => px.coord > f.get('start') &&
        px.coord < f.get('end') &&
        px.refName === assembly.getCanonicalRefName(f.get('refName')));
    return (react_1.default.createElement(material_1.Tooltip, { open: open, placement: "top", title: [(0, util_1.stringify)(px), cytoband === null || cytoband === void 0 ? void 0 : cytoband.get('name')].join(' '), arrow: true },
        react_1.default.createElement("div", { className: classes.guide, style: { left: guideX } })));
});
const OverviewRubberband = (0, mobx_react_1.observer)(function OverviewRubberband({ model, overview, ControlComponent = react_1.default.createElement("div", null), }) {
    const { cytobandOffset } = model;
    const [startX, setStartX] = (0, react_1.useState)();
    const [currentX, setCurrentX] = (0, react_1.useState)();
    const [guideX, setGuideX] = (0, react_1.useState)();
    const controlsRef = (0, react_1.useRef)(null);
    const { classes } = useStyles();
    const mouseDragging = startX !== undefined;
    (0, react_1.useEffect)(() => {
        function globalMouseMove(event) {
            const ref = controlsRef.current;
            if (ref && mouseDragging) {
                setCurrentX((0, util_2.getRelativeX)(event, ref));
            }
        }
        function globalMouseUp() {
            // click and drag
            if (startX !== undefined && currentX !== undefined) {
                if (Math.abs(currentX - startX) > 3) {
                    const left = Math.min(startX, currentX);
                    const right = Math.max(startX, currentX);
                    model.moveTo(overview.pxToBp(left - cytobandOffset), overview.pxToBp(right - cytobandOffset));
                }
            }
            // just a click
            if (startX !== undefined && currentX === undefined) {
                const click = overview.pxToBp(startX - cytobandOffset);
                if (!click.refName) {
                    (0, util_1.getSession)(model).notify('unknown position clicked');
                    console.error('unknown position clicked', click);
                }
                else {
                    model.centerAt(Math.round(click.coord), click.refName, click.index);
                }
            }
            setStartX(undefined);
            setCurrentX(undefined);
            if (startX !== undefined) {
                setGuideX(undefined);
            }
        }
        function globalKeyDown(event) {
            if (event.key === 'Escape') {
                setStartX(undefined);
                setCurrentX(undefined);
            }
        }
        if (mouseDragging) {
            window.addEventListener('mousemove', globalMouseMove, true);
            window.addEventListener('mouseup', globalMouseUp, true);
            window.addEventListener('keydown', globalKeyDown, true);
            return () => {
                window.removeEventListener('mousemove', globalMouseMove, true);
                window.removeEventListener('mouseup', globalMouseUp, true);
                window.removeEventListener('keydown', globalKeyDown, true);
            };
        }
        return () => { };
    }, [mouseDragging, currentX, startX, model, overview, cytobandOffset]);
    function mouseDown(event) {
        event.preventDefault();
        event.stopPropagation();
        setStartX((0, util_2.getRelativeX)(event, controlsRef.current));
    }
    function mouseMove(event) {
        setGuideX((0, util_2.getRelativeX)(event, controlsRef.current));
    }
    function mouseOut() {
        setGuideX(undefined);
    }
    if (startX === undefined) {
        return (react_1.default.createElement("div", { className: classes.rel },
            guideX !== undefined ? (react_1.default.createElement(HoverTooltip, { model: model, open: !mouseDragging, overview: overview, guideX: guideX })) : null,
            react_1.default.createElement("div", { className: classes.rubberbandControl, ref: controlsRef, onMouseDown: mouseDown, onMouseOut: mouseOut, onMouseMove: mouseMove }, ControlComponent)));
    }
    let left = startX || 0;
    let width = 0;
    if (currentX !== undefined) {
        left = Math.min(currentX, startX);
        width = currentX - startX;
    }
    // calculate the start and end bp of drag
    let leftBpOffset;
    let rightBpOffset;
    if (startX) {
        leftBpOffset = overview.pxToBp(startX - cytobandOffset);
        rightBpOffset = overview.pxToBp(startX + width - cytobandOffset);
        if (currentX !== undefined && currentX < startX) {
            ;
            [leftBpOffset, rightBpOffset] = [rightBpOffset, leftBpOffset];
        }
    }
    return (react_1.default.createElement("div", { className: classes.rel },
        leftBpOffset && rightBpOffset ? (react_1.default.createElement(RubberbandSpan_1.default, { leftBpOffset: leftBpOffset, rightBpOffset: rightBpOffset, width: Math.abs(width), left: left })) : null,
        react_1.default.createElement("div", { "data-testid": "rubberband_controls", className: classes.rubberbandControl, ref: controlsRef, onMouseDown: mouseDown, onMouseOut: mouseOut, onMouseMove: mouseMove }, ControlComponent)));
});
exports.default = OverviewRubberband;
