// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/OverviewHighlight.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const mobx_react_1 = require("mobx-react");
const mui_1 = require("tss-react/mui");
const colord_1 = require("@jbrowse/core/util/colord");
const util_1 = require("@jbrowse/core/util");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    highlight: {
        height: '100%',
        position: 'absolute',
        background: (0, colord_1.colord)(theme.palette.highlight.main).alpha(0.35).toRgbString(),
        borderLeft: `1px solid ${theme.palette.highlight.main}`,
        borderRight: `1px solid ${theme.palette.highlight.main}`,
    },
}));
const OverviewHighlight = (0, mobx_react_1.observer)(function OverviewHighlight({ model, overview, }) {
    const { classes } = useStyles();
    const { highlight, cytobandOffset } = model;
    const session = (0, util_1.getSession)(model);
    const { assemblyManager } = session;
    return highlight
        .map(r => {
        var _a;
        const asm = assemblyManager.get(r.assemblyName);
        const refName = (_a = asm === null || asm === void 0 ? void 0 : asm.getCanonicalRefName(r.refName)) !== null && _a !== void 0 ? _a : r.refName;
        const s = overview.bpToPx({
            ...r,
            refName,
            coord: r.start,
        });
        const e = overview.bpToPx({
            ...r,
            refName,
            coord: r.end,
        });
        return s !== undefined && e !== undefined
            ? {
                width: Math.abs(e - s),
                left: s + cytobandOffset,
            }
            : undefined;
    })
        .filter(util_1.notEmpty)
        .map(({ left, width }, idx) => (react_1.default.createElement("div", { 
        /* biome-ignore lint/suspicious/noArrayIndexKey: */
        key: `${left}_${width}_${idx}`, className: classes.highlight, style: {
            width: width,
            left: left,
        } })));
});
exports.default = OverviewHighlight;
