// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/ScalebarRefNameLabels.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    refLabel: {
        fontSize: 11,
        position: 'absolute',
        left: 2,
        top: -1,
        fontWeight: 'bold',
        lineHeight: 'normal',
        zIndex: 1,
        background: theme.palette.background.paper,
    },
    b0: {
        left: 0,
        zIndex: 100,
    },
}));
const ScalebarRefNameLabels = (0, mobx_react_1.observer)(function ({ model }) {
    const { classes, cx } = useStyles();
    const { staticBlocks, offsetPx, scaleBarDisplayPrefix } = model;
    // find the block that needs pinning to the left side for context
    let lastLeftBlock = 0;
    staticBlocks.forEach((block, i) => {
        if (block.offsetPx - offsetPx < 0) {
            lastLeftBlock = i;
        }
    });
    const val = scaleBarDisplayPrefix();
    const b0 = staticBlocks.blocks[0];
    return (react_1.default.createElement(react_1.default.Fragment, null,
        (b0 === null || b0 === void 0 ? void 0 : b0.type) !== 'ContentBlock' && val ? (react_1.default.createElement(material_1.Typography, { className: cx(classes.b0, classes.refLabel) }, val)) : null,
        staticBlocks.map((block, index) => {
            const { offsetPx: blockOffsetPx, isLeftEndOfDisplayedRegion, key, type, refName, } = block;
            const last = index === lastLeftBlock;
            return type === 'ContentBlock' &&
                (isLeftEndOfDisplayedRegion || last) ? (react_1.default.createElement(material_1.Typography, { key: `refLabel-${key}-${index}`, style: {
                    left: last
                        ? Math.max(0, -offsetPx)
                        : blockOffsetPx - offsetPx - 1,
                    paddingLeft: last ? 0 : 1,
                }, className: classes.refLabel, "data-testid": `refLabel-${refName}` },
                last && val ? `${val}:` : '',
                refName)) : null;
        })));
});
exports.default = ScalebarRefNameLabels;
