// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/ScalebarCoordinateTicks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const react_1 = __importDefault(require("react"));
const util_1 = require("@jbrowse/core/util");
// locals
const Block_1 = require("../../BaseLinearDisplay/components/Block");
const util_2 = require("../util");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    majorTickLabel: {
        fontSize: 11,
        zIndex: 1,
        background: theme.palette.background.paper,
        lineHeight: 'normal',
        pointerEvents: 'none',
    },
    tick: {
        position: 'absolute',
        width: 0,
        display: 'flex',
        justifyContent: 'center',
        pointerEvents: 'none',
    },
}));
const ScalebarCoordinateTicks = function ({ block, bpPerPx, }) {
    const { classes } = useStyles();
    const { reversed, start, end } = block;
    const ticks = (0, util_2.makeTicks)(start, end, bpPerPx, true, false);
    return (react_1.default.createElement(Block_1.ContentBlock, { block: block }, ticks.map(({ type, base }) => {
        if (type === 'major') {
            const x = (reversed ? end - base : base - start) / bpPerPx;
            const baseNumber = base + 1;
            return (react_1.default.createElement("div", { key: base, className: classes.tick, style: { left: x } }, baseNumber ? (react_1.default.createElement(material_1.Typography, { className: classes.majorTickLabel }, (0, util_1.getTickDisplayStr)(baseNumber, bpPerPx))) : null));
        }
        return null;
    })));
};
exports.default = ScalebarCoordinateTicks;
