// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/TrackLabelDragHandle.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const mui_1 = require("tss-react/mui");
// icons
const DragIndicator_1 = __importDefault(require("@mui/icons-material/DragIndicator"));
const useStyles = (0, mui_1.makeStyles)()({
    dragHandle: {
        cursor: 'grab',
    },
    dragHandleIcon: {
        display: 'inline-block',
        verticalAlign: 'middle',
        pointerEvents: 'none',
    },
});
function TrackLabelDragHandle({ trackId, view, track, }) {
    const { classes } = useStyles();
    return (react_1.default.createElement("span", { draggable: true, className: classes.dragHandle, onDragStart: event => {
            const target = event.currentTarget;
            if (target.parentNode) {
                const parent = target.parentNode;
                event.dataTransfer.setDragImage(parent, 20, 20);
                view.setDraggingTrackId(track.id);
            }
        }, onDragEnd: () => {
            view.setDraggingTrackId(undefined);
        }, "data-testid": `dragHandle-${view.id}-${trackId}` },
        react_1.default.createElement(DragIndicator_1.default, { className: classes.dragHandleIcon, fontSize: "small" })));
}
exports.default = TrackLabelDragHandle;
