// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/TrackLabelContainer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
const TrackLabel_1 = __importDefault(require("./TrackLabel"));
const useStyles = (0, mui_1.makeStyles)()({
    trackLabel: {
        zIndex: 3,
    },
    trackLabelOffset: {
        position: 'relative',
        display: 'inline-block',
    },
    trackLabelOverlap: {
        position: 'absolute',
    },
});
const TrackLabelContainer = (0, mobx_react_1.observer)(function ({ track, view, }) {
    const { classes, cx } = useStyles();
    const display = track.displays[0];
    const { trackLabel, trackLabelOverlap, trackLabelOffset } = classes;
    const labelStyle = view.trackLabelsSetting !== 'overlapping' || display.prefersOffset
        ? trackLabelOffset
        : trackLabelOverlap;
    return view.trackLabelsSetting !== 'hidden' ? (react_1.default.createElement(TrackLabel_1.default, { track: track, className: cx(trackLabel, labelStyle) })) : null;
});
exports.default = TrackLabelContainer;
