// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/VerticalGuide.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
const util_1 = require("@jbrowse/core/util");
const useStyles = (0, mui_1.makeStyles)()({
    guide: {
        pointerEvents: 'none',
        height: '100%',
        width: 1,
        position: 'absolute',
        zIndex: 10,
    },
});
const VerticalGuide = (0, mobx_react_1.observer)(function VerticalGuide({ model, coordX, }) {
    const { classes } = useStyles();
    return (react_1.default.createElement(material_1.Tooltip, { open: true, placement: "top", title: (0, util_1.stringify)(model.pxToBp(coordX)), arrow: true },
        react_1.default.createElement("div", { className: classes.guide, style: {
                left: coordX,
                background: 'red',
            } })));
});
exports.default = VerticalGuide;
