// @jbrowse/plugin-linear-genome-view/dist/BaseLinearDisplay/models/TrackHeightMixin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = TrackHeightMixin;
const configuration_1 = require("@jbrowse/core/configuration");
const mobx_state_tree_1 = require("mobx-state-tree");
const minDisplayHeight = 20;
/**
 * #stateModel TrackHeightMixin
 * #category display
 */
function TrackHeightMixin() {
    return mobx_state_tree_1.types
        .model({
        /**
         * #property
         */
        heightPreConfig: mobx_state_tree_1.types.maybe(mobx_state_tree_1.types.refinement('displayHeight', mobx_state_tree_1.types.number, n => n >= minDisplayHeight)),
    })
        .volatile(() => ({
        /**
         * #property
         */
        scrollTop: 0,
    }))
        .views(self => ({
        get height() {
            var _a;
            // @ts-expect-error
            return (_a = self.heightPreConfig) !== null && _a !== void 0 ? _a : (0, configuration_1.getConf)(self, 'height');
        },
    }))
        .actions(self => ({
        /**
         * #action
         */
        setScrollTop(scrollTop) {
            self.scrollTop = scrollTop;
        },
        /**
         * #action
         */
        setHeight(displayHeight) {
            self.heightPreConfig = Math.max(displayHeight, minDisplayHeight);
            return self.height;
        },
        /**
         * #action
         */
        resizeHeight(distance) {
            const oldHeight = self.height;
            const newHeight = this.setHeight(self.height + distance);
            return newHeight - oldHeight;
        },
    }));
}
