// @jbrowse/plugin-linear-genome-view/dist/LinearBasicDisplay/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configSchema = exports.modelFactory = void 0;
exports.default = LinearBasicDisplay;
const pluggableElementTypes_1 = require("@jbrowse/core/pluggableElementTypes");
const BaseLinearDisplay_1 = require("../BaseLinearDisplay");
// locals
const configSchema_1 = __importDefault(require("./configSchema"));
const model_1 = __importDefault(require("./model"));
function LinearBasicDisplay(pluginManager) {
    pluginManager.addDisplayType(() => {
        const config = (0, configSchema_1.default)(pluginManager);
        return new pluggableElementTypes_1.DisplayType({
            name: 'LinearBasicDisplay',
            displayName: 'Basic feature display',
            configSchema: config,
            stateModel: (0, model_1.default)(config),
            trackType: 'FeatureTrack',
            viewType: 'LinearGenomeView',
            ReactComponent: BaseLinearDisplay_1.BaseLinearDisplayComponent,
        });
    });
}
var model_2 = require("./model");
Object.defineProperty(exports, "modelFactory", { enumerable: true, get: function () { return __importDefault(model_2).default; } });
var configSchema_2 = require("./configSchema");
Object.defineProperty(exports, "configSchema", { enumerable: true, get: function () { return __importDefault(configSchema_2).default; } });
