// @jbrowse/plugin-linear-genome-view/dist/LinearBareDisplay/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stateModelFactory = exports.configSchemaFactory = void 0;
exports.default = LinearBareDisplayF;
const pluggableElementTypes_1 = require("@jbrowse/core/pluggableElementTypes");
// locals
const configSchema_1 = require("./configSchema");
const model_1 = require("./model");
const BaseLinearDisplay_1 = require("../BaseLinearDisplay/");
function LinearBareDisplayF(pluginManager) {
    pluginManager.addDisplayType(() => {
        const configSchema = (0, configSchema_1.configSchemaFactory)(pluginManager);
        return new pluggableElementTypes_1.DisplayType({
            name: 'LinearBareDisplay',
            configSchema,
            displayName: 'Bare feature display',
            stateModel: (0, model_1.stateModelFactory)(configSchema),
            trackType: 'BasicTrack',
            viewType: 'LinearGenomeView',
            ReactComponent: BaseLinearDisplay_1.BaseLinearDisplayComponent,
        });
    });
}
var configSchema_2 = require("./configSchema");
Object.defineProperty(exports, "configSchemaFactory", { enumerable: true, get: function () { return configSchema_2.configSchemaFactory; } });
var model_2 = require("./model");
Object.defineProperty(exports, "stateModelFactory", { enumerable: true, get: function () { return model_2.stateModelFactory; } });
