// @jbrowse/plugin-linear-genome-view/dist/LinearBareDisplay/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stateModelFactory = stateModelFactory;
const configuration_1 = require("@jbrowse/core/configuration");
const tracks_1 = require("@jbrowse/core/util/tracks");
const mobx_state_tree_1 = require("mobx-state-tree");
const BaseLinearDisplay_1 = require("../BaseLinearDisplay");
/**
 * #stateModel LinearBareDisplay
 * #category display
 * extends
 * - [BaseLinearDisplay](../baselineardisplay)
 */
function stateModelFactory(configSchema) {
    return mobx_state_tree_1.types
        .compose('LinearBareDisplay', BaseLinearDisplay_1.BaseLinearDisplay, mobx_state_tree_1.types.model({
        /**
         * #property
         */
        type: mobx_state_tree_1.types.literal('LinearBareDisplay'),
        /**
         * #property
         */
        configuration: (0, configuration_1.ConfigurationReference)(configSchema),
    }))
        .views(self => {
        const { renderProps: superRenderProps } = self;
        return {
            /**
             * #method
             */
            renderProps() {
                return {
                    ...superRenderProps(),
                    ...(0, tracks_1.getParentRenderProps)(self),
                    rpcDriverName: self.rpcDriverName,
                    config: self.configuration.renderer,
                };
            },
            /**
             * #getter
             */
            get rendererTypeName() {
                return self.configuration.renderer.type;
            },
        };
    });
}
