// @jbrowse/core/BaseFeatureWidget/BaseFeatureDetail/BaseCard.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BaseCard;
const react_1 = __importDefault(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
// icons
const ExpandMore_1 = __importDefault(require("@mui/icons-material/ExpandMore"));
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    expansionPanelDetails: {
        display: 'block',
        padding: theme.spacing(1),
    },
    icon: {
        color: theme.palette.tertiary.contrastText || '#fff',
    },
}));
function BaseCard({ children, title, defaultExpanded = true, }) {
    const { classes } = useStyles();
    return (react_1.default.createElement(material_1.Accordion, { defaultExpanded: defaultExpanded },
        react_1.default.createElement(material_1.AccordionSummary, { expandIcon: react_1.default.createElement(ExpandMore_1.default, { className: classes.icon }) },
            react_1.default.createElement(material_1.Typography, { variant: "button" }, title)),
        react_1.default.createElement(material_1.AccordionDetails, { className: classes.expansionPanelDetails }, children)));
}
